src/VisualStudio/Core/Def/DocumentOutline/DocumentSymbolDataViewModel.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.ComponentModel;
+using System.Runtime.CompilerServices;
+using Microsoft.CodeAnalysis.Editor.Wpf;
+using Microsoft.VisualStudio.Imaging.Interop;
+using Microsoft.VisualStudio.Text;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    using SymbolKind = LanguageServer.Protocol.SymbolKind;
+
+    /// <summary>
+    /// A ViewModel over <see cref="DocumentSymbolData"/>. The only items that are mutable on this type are <see
+    /// cref="IsExpanded"/> and <see cref="IsSelected"/>. It is expected that these can be modified from any thread with
+    /// INotifyPropertyChanged notifications being marshalled to the correct thread by WPF if there needs to be a change
+    /// to the visual presentation.
+    /// </summary>
+    internal sealed class DocumentSymbolDataViewModel : INotifyPropertyChanged, IEquatable<DocumentSymbolDataViewModel>
+    {
+        public DocumentSymbolData Data { get; }
+        public ImmutableArray<DocumentSymbolDataViewModel> Children { get; }
+
+        /// <summary>
+        /// Necessary because we cannot convert to this type dynamically in WPF.
+        /// </summary>
+        public ImageMoniker ImageMoniker => Data.Glyph.GetImageMoniker();
+
+        private bool _isExpanded = true;
+        private bool _isSelected = false;
+
+        public bool IsExpanded
+        {
+            get => _isExpanded;
+            set => SetProperty(ref _isExpanded, value);
+        }
+
+        public bool IsSelected
+        {
+            get => _isSelected;
+            set => SetProperty(ref _isSelected, value);
+        }
+
+        public DocumentSymbolDataViewModel(
+            DocumentSymbolData data,
+            ImmutableArray<DocumentSymbolDataViewModel> children)
+        {
+            Data = data;
+            Children = children;
+        }
+
+        private static readonly PropertyChangedEventArgs _isExpandedPropertyChangedEventArgs = new PropertyChangedEventArgs(nameof(IsExpanded));
```

Comment:
Is there a real reason to make static instances here? I wouldn't expect it to be too costly without them, but maybe?

Reply:
@jmarolf ?

Reply:
In a file with thousands of members we would be creating many instances of these unnecessarily.

Reply:
Is that because the expansion state is being automatically changed somewhere?